pallets/gear/src/manager/task.rs

```diff
@@ -31,24 +31,70 @@ use common::{
     storage::*,
     Origin, PausedProgramStorage, Program, ProgramStorage,
 };
+use core::cmp;
 use frame_support::traits::{Currency, ExistenceRequirement};
 use gear_core::{
+    code::MAX_WASM_PAGE_COUNT,
     ids::{CodeId, MessageId, ProgramId, ReservationId},
     message::{DispatchKind, ReplyMessage},
+    pages::{GEAR_PAGE_SIZE, WASM_PAGE_SIZE},
 };
 use gear_core_errors::{ErrorReplyReason, SignalCode};
 use sp_runtime::traits::Zero;
 
+pub fn get_maximum_task_gas<T: Config>(task: &ScheduledTask<T::AccountId>) -> Gas {
+    use ScheduledTask::*;
+
+    match task {
+        PauseProgram(_) => {
+            let count =
+                u32::from(MAX_WASM_PAGE_COUNT * (WASM_PAGE_SIZE / GEAR_PAGE_SIZE) as u16 / 2);
+            cmp::max(
+                <T as Config>::WeightInfo::tasks_pause_program(count).ref_time(),
+                <T as Config>::WeightInfo::tasks_pause_program_uninited(count).ref_time(),
+            )
+        }
+        RemoveCode(_) => todo!("#646"),
+        RemoveFromMailbox(_, _) => {
+            <T as Config>::WeightInfo::tasks_remove_from_mailbox().ref_time()
+        }
+        RemoveFromWaitlist(_, _) => {
+            <T as Config>::WeightInfo::tasks_remove_from_waitlist().ref_time()
+        }
+        RemovePausedProgram(_) => todo!("#646"),
+        WakeMessage(_, _) => cmp::max(
+            <T as Config>::WeightInfo::tasks_wake_message().ref_time(),
+            <T as Config>::WeightInfo::tasks_wake_message_no_wake().ref_time(),
+        ),
+        SendDispatch(_) => <T as Config>::WeightInfo::tasks_send_dispatch().ref_time(),
+        SendUserMessage { .. } => cmp::max(
+            <T as Config>::WeightInfo::tasks_send_user_message_to_mailbox().ref_time(),
+            <T as Config>::WeightInfo::tasks_send_user_message().ref_time(),
+        ),
+        RemoveGasReservation(_, _) => {
+            <T as Config>::WeightInfo::tasks_remove_gas_reservation().ref_time()
+        }
+        RemoveResumeSession(_) => {
+            <T as Config>::WeightInfo::tasks_remove_resume_session().ref_time()
+        }
+    }
+}
+
 impl<T: Config> TaskHandler<T::AccountId> for ExtManager<T>
 where
     T::AccountId: Origin,
 {
-    fn pause_program(&mut self, program_id: ProgramId) {
+    fn pause_program(&mut self, program_id: ProgramId) -> Gas {
         log::debug!("pause_program; id = {:?}", program_id);
 
         let program = ProgramStorageOf::<T>::get_program(program_id)
             .unwrap_or_else(|| unreachable!("Program to pause not found."));
 
+        let pages_with_data = match program {
+            Program::Active(ref p) => p.pages_with_data.len() as u32,
```

Comment:
Consider introduce `Program::as_active`

Reply:
There is an implementation of TryFrom for ActiveProgram, so will rewrite code using it.